src/main/java/software/amazon/encryption/s3/internal/InstructionFileConfig.java

```diff
@@ -0,0 +1,103 @@
+    final private S3AsyncClient _s3AsyncClient;
+    final private S3Client _s3Client;
+
+    private InstructionFileConfig(final Builder builder) {
+        _clientType = builder._clientType;
+        _disableInstructionFile = builder._disableInstructionFile;
+        _s3Client = builder._s3Client;
+        _s3AsyncClient = builder._s3AsyncClient;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public enum InstructionFileClientType {
+        SYNCHRONOUS,
+        ASYNC
+    }
+
+    ResponseInputStream<GetObjectResponse> getInstructionFile(GetObjectRequest request) {
+        if (_disableInstructionFile) {
+            throw new S3EncryptionClientException("Instruction File has been disabled!");
+        }
+        switch (_clientType) {
+            case SYNCHRONOUS:
+                return _s3Client.getObject(request);
+            case ASYNC:
+                return _s3AsyncClient.getObject(request, AsyncResponseTransformer.toBlockingInputStream()).join();
+            default:
+                throw new S3EncryptionClientException("Unknown Instruction File Type");
+        }
+    }
+
+    /**
+     * Closes the S3Client or S3AsyncClient used for instruction files.
+     */
+    public void closeClient() {
+        if (_s3AsyncClient != null) {
+            _s3AsyncClient.close();
+        }
+        if (_s3Client != null) {
+            _s3Client.close();
+        }
+    }
+
+    public static class Builder {
+        private InstructionFileClientType _clientType;
+        private boolean _disableInstructionFile;
+        private S3AsyncClient _s3AsyncClient;
+        private S3Client _s3Client;
+
+        /**
+         * When set to true, the S3 Encryption Client will not attempt to get instruction files.
+         * @param disableInstructionFile
+         * @return
+         */
+        public Builder disableInstructionFile(boolean disableInstructionFile) {
+            _disableInstructionFile = disableInstructionFile;
+            return this;
+        }
+
+        public Builder instructionFileClient(S3Client instructionFileClient) {
+            _s3Client = instructionFileClient;
+            return this;
+        }
+
+        public Builder instructionFileAsyncClient(S3AsyncClient instructionFileAsyncClient) {
+            _s3AsyncClient = instructionFileAsyncClient;
+            return this;
+        }
+        public InstructionFileConfig build() {
+            if (_s3Client != null && _s3AsyncClient != null) {
+                throw new S3EncryptionClientException("Only one instruction file client may be set.");
+            }
+            if (_s3Client != null) {
+                _clientType = InstructionFileClientType.SYNCHRONOUS;
+            } else if (_s3AsyncClient != null){
+                _clientType = InstructionFileClientType.ASYNC;
+            } else {
+                throw new S3EncryptionClientException("At least one instruction file client must be set.");
+            }
```

Comment:
This class is not correct.
In particular, if `_disableInstructionFile` is true,
it should be fine for there to be no client.